app/views/buyers/applications/new.html.erb

```diff
@@ -14,10 +14,12 @@
 
   <%= form.semantic_errors %>
   <%= form.inputs do %>
-    <%= form.input :plan, :as => :plan_selector,
-                          :collection => @plans, :include_blank => false,
-                          :label => ApplicationPlan.model_name.human,
-                          :default_plan => @plans.respond_to?(:default) ? @plans.default : @plans.first %>
+    <li>
+      <label>Application plan</label>
+      <%# FIXME: remove do/end here, for some reason it's required %>
+      <%= content_tag :span, id:'plan_selector', plans: @plans.to_json do %>
```

Comment:
So, this used to be progressive enhanced form and work even without javascript.

IMO that is the correct way to properly replace components, render their html and then transform them into different component with JS.

Now if there is an error with the inline javascript, or some other breakage this form will be unusable.

Reply:
Agreed but to be fair the page as a whole was already not working without javascript. And 3scale as a whole is pretty useless without javascript enabled. Not saying that's a good thing but stating it as a fact.

Moving forward, PatternFly4 doesn't seem to care about progressive enhancement. And React as a library is not pushing people in that direction either. Doesn't mean we shouldn't but will be a swim against the current.

Reply:
I agree that 3scale as a whole is pretty unusable without javascript. But there are many parts where it just works.
Indeed, the app creation already required JS, as we can see in the cuke: https://github.com/3scale/porta/blob/6f74cf39b5bfef598ef24388b17fb98ad3618484/features/old/applications/providers/create_application.feature

My point is that there is a bunch of inline javascript everywhere, that is fairly easy to break. It is not unit tested, only tested by cukes with JS enabled. It is easy to have a bug that will render this whole page unusable.

If we can use progressive enhancement, IMO we should try to, because it is safer and offers us a working page, if something fails. Not necessarily to make the entire page usable without javascript.

Also, running all the test suite only with javascript would significantly increase the build times, so that is something we should keep in mind and try to keep it functioning even without JS.

Reply:
I agree that progressive enhancement is the right way to go about web development. I will bring it up in system team.

Reply:
I think we could simply show a warning "please enable Javascript" ... nowadays this is a very unlike scenario, only perhaps for a limited connectivity environment such as mobile..

Reply:
Do you want to show a warning "please enable JavaScript" even when it is enabled and there is runtime error in the javascript code?

Reply:
Also we should minimize the JS injected in Rails views, and having these modules fully tested in JavaScript. We could start building pieces in HTML or Rails helpers and then replacing them when JS is loaded. Kind of a PITA sometimes hehehe, but worth discussing it